wallet-api/src/Wallet/Emulator/Types.hs

```diff
@@ -408,23 +418,41 @@ data ValidatedBlock = ValidatedBlock
     , vlbRest   :: [Tx]
     -- ^ Transactions that haven't been validated because the current slot is
     --   not in their validation interval
+    , vlbIndex  :: Index.UtxoIndex
+    -- ^ Update UTXO index
     }
 
--- | Validate a block in an [[EmulatorState]], returning the valid transactions
---   and all success/failure events
-validateBlock :: EmulatorState -> [Tx] -> ValidatedBlock
-validateBlock emState txns = ValidatedBlock block events rest where
-    (eligibleTxns, rest) = partition canValidateNow txns
-    processed = (\tx -> (tx, validateEm emState tx)) <$> eligibleTxns
-    validTxns = fst <$> filter (isNothing . snd) processed
-    block = validTxns
-    mkEvent (t, result) =
-        case result of
-            Nothing  -> TxnValidate (hashTx t)
-            Just err -> TxnValidationFail (hashTx t) err
-    events = mkEvent <$> processed
-    canValidateNow tx = $$(Interval.member) currentSlot (txValidRange tx)
-    currentSlot = emState ^. chainNewestFirst . to lastSlot
+-- | Validate a block given the current slot and UTXO set, returning the valid 
+--   transactions, success/failure events, remaining transactions and the 
+--   updated UTXO set.
+validateBlock :: Slot -> Index.UtxoIndex -> [Tx] -> ValidatedBlock
+validateBlock currentSlot idx txns = 
+    let
+        -- Select those transactions that can be validated in the 
+        -- current slot
+        (eligibleTxns, rest) = partition (canValidateNow currentSlot) txns
+
+        -- Validate eligible transactions, updating the UTXO index each time
+        (processed, idx') = runState (traverse (\t -> fmap (t,) (validateEm currentSlot t)) eligibleTxns) idx
```

Comment:
This is a bit fancy. I'd probably write a do-block as the argument to the `runState`.

Reply:
done